cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java

```diff
@@ -363,25 +381,77 @@ String removeAppVersion(String scheduleId) {
   /**
    * Method to add version in StreamSizeSchedule row key in SchedulerStore.
    *
-   * @throws Exception
+   * @throws InterruptedException
+   * @throws IOException
+   * @throws DatasetManagementException
    */
-  public void upgrade()
-    throws InterruptedException, TransactionFailureException, IOException, DatasetManagementException {
-    initialize();
-    factory.createExecutor(ImmutableList.of((TransactionAware) table))
-      .execute(new TransactionExecutor.Subroutine() {
-        @Override
-        public void apply() {
-          upgradeVersionKeys();
+  public void upgrade() throws InterruptedException, IOException, DatasetManagementException {
+    // Wait until the store is initialized
+    while (!storeInitialized.get()) {
+      TimeUnit.SECONDS.sleep(10);
+    }
+
+    if (isUpgradeComplete()) {
+      LOG.info("{} is already upgraded.", NAME);
+      return;
+    }
+
+    final AtomicInteger maxNumberUpdateRows = new AtomicInteger(1000);
+    final AtomicInteger sleepTimeInMins = new AtomicInteger(1);
+    LOG.info("Starting upgrade of {}.", NAME);
+    while (!isUpgradeComplete()) {
```

Comment:
can have a local variable `final AtomicBoolean upgradeComplete` to be set true directly after `table.put(APP_VERSION_UPGRADE_KEY, COLUMN, Bytes.toBytes(ProjectInfo.getVersion().toString()));` and check the local variable `upgradeComplete` here instead of calling `isUpgradeComplete()`, which may not be updated immediately